controllers/cloud.redhat.com/clowdenvironment_reconciliation.go

```diff
@@ -0,0 +1,489 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	crd "github.com/RedHatInsights/clowder/apis/cloud.redhat.com/v1alpha1"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/clowderconfig"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/errors"
+	"github.com/RedHatInsights/clowder/controllers/cloud.redhat.com/providers"
+	rc "github.com/RedHatInsights/rhc-osdk-utils/resource_cache"
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	cond "sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	SKIPRECONCILE = "SKIPRECONCILE"
+)
+
+//During reconciliation we handle errors in 2 ways: sometimes we want to error out of reconciliation and sometimes we want to skip reconciliation.
+func shouldSkipReconciliation(err error) bool {
+	if err != nil && err.Error() == SKIPRECONCILE {
+		return true
+	}
+	return false
+}
+
+func MakeClowdEnvReconciliation(cache *rc.ObjectCache,
+	recorder record.EventRecorder,
+	ctx *context.Context,
+	client client.Client,
+	env *crd.ClowdEnvironment,
+	log *logr.Logger) ClowdEnvironmentReconciliation {
+	return ClowdEnvironmentReconciliation{
+		cache:    cache,
+		recorder: recorder,
+		ctx:      ctx,
+		client:   client,
+		env:      env,
+		log:      log,
+	}
+}
+
+//ClowdEnvironmentReconciliation represents a single reconciliation event
+//This type is created by ClowdEnvironmentReconciler which handles the reconciliation cycle as a whole
+//ClowdEnvironmentReconciliation encapsulates all of the state and logic requires for a single
+//reconciliation event
+type ClowdEnvironmentReconciliation struct {
+	cache    *rc.ObjectCache
+	recorder record.EventRecorder
+	ctx      *context.Context
+	client   client.Client
+	env      *crd.ClowdEnvironment
+	log      *logr.Logger
+}
+
+//Returns a list of step methods that should be run during reconciliation
+func (r *ClowdEnvironmentReconciliation) steps() []func() (ctrl.Result, error) {
+	return []func() (ctrl.Result, error){
+		r.markedForDeletion,
+		r.addFinalizerIfRequired,
+		r.perProviderMetrics,
+		r.setToBeDisabled,
+		r.initTargetNamespace,
+		r.isTargetNamespaceMarkedForDeletion,
+		r.runProviders,
+		r.applyCache,
+		r.setAppInfo,
+		r.setEnvResourceStatus,
+		r.setPrometheusStatus,
+		r.setEnvStatus,
+		r.finalStatusError,
+		r.deleteUnusedResources,
+		r.setClowdEnvConditions,
+		r.logSuccess,
+	}
+}
+
+//Public method to iterate through the steps of the reconciliation process
+func (r *ClowdEnvironmentReconciliation) Reconcile() (ctrl.Result, error) {
+	r.log.Info("Reconciliation started")
+	//The env stays locked for the entire reconciliation
```

Comment:
Not a problem with this change

Reply:
Awesome! I was super nervous about this.